### PR TITLE
feat: ms-appdata URI file path evaluator

### LINKED ISSUE #2925: ms-appdata:/// image paths are treated as Resource
## Current behavior

When an image is loaded with the `ms-appdata:///` URI, for example:

```
ms-appdata:///local/icon.png
```

The path is treated as an app package _resource_, although it should in fact point to a local file.

## Expected behavior

Images stored in `ms-appdata:///` folders are loaded from their proper location.

## How to reproduce it (as minimally and precisely as possible)

Put an image into `ApplicationData.Current.LocalFolder` and try to load it in an image control.

## Environment

Nuget Package: 2.2 dev

Package Version(s): 2.2. dev

Affected platform(s):

- [x] iOS
- [x] Android
- [ ] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [x] macOS
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

### PR DESCRIPTION
**Pull request by MartinZikmund**
GitHub Issue (If applicable): fixes #2925 
## PR Type

What kind of change does this PR introduce?

- Feature

## What is the current behavior?

ms-appdata: URI is evaluated as a package resource.

## What is the new behavior?

ms-appdata: URI evaluates to a corresponding application data file path.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

### CODE REVIEW

**Comment by pkar70** [comment]
So, maybe adding `Windows.Storage.StorageFile.GetFileFromApplicationUriAsync(uri);` can be done?

**Comment by MartinZikmund** [comment]
@pkar70 Yes, the `ms-appdata:` URI can be solved using the evaluator. `ms-appx:` is a bit more complicated, as I am not sure if both AndroidAsset and BundleResource can be represented as a "file path"

**Comment by pkar70** [comment]
One of my app, converted from VB/UWP to Uno, differs between Android and UWP version: Android version has some functionality stripped down.

            Uri oPicUri = new Uri("ms-appx:///Assets/" + iRok + ".gif");
            Windows.Storage.StorageFile oFile;
            try
            {   
                oFile = await Windows.Storage.StorageFile.GetFileFromApplicationUriAsync(oPicUri);
            }
            catch
            {
                return false;
            }

            Windows.UI.Xaml.Media.Imaging.BitmapImage oBmp = new Windows.UI.Xaml.Media.Imaging.BitmapImage();
            oBmp.UriSource = oPicUri;
            uiPic.Source = oBmp;

So, if `Windows.Storage.StorageFile.GetFileFromApplicationUriAsync(oPicUri);` would work, I can re-add this functionality to Android version....

File: src/Uno.Foundation/UriExtensions.cs [hunk 1]
```diff
@@ -57,10 +57,24 @@ public static string GetExtension(this Uri uri)
...
+			{
+				throw new ArgumentNullException(nameof(uri));
+			}
+
+			return uri.Scheme.Equals("ms-appdata", StringComparison.OrdinalIgnoreCase);
+		}
+
 		public static bool IsLocalResource(this Uri uri)
 		{
-			return uri.Scheme.Equals("ms-appx", StringComparison.OrdinalIgnoreCase)
```
Comment (jeromelaban) [inline src/Uno.Foundation/UriExtensions.cs:62]: This is a breaking change, but considering it was not supported properly, that's likely fine.

File: src/Uno.Foundation/UriExtensions.cs [hunk 2]
```diff
@@ -57,10 +57,24 @@ public static string GetExtension(this Uri uri)
 			return String.Empty;
 		}
 
+		public static bool IsAppData(this Uri uri)
```
Comment (jeromelaban) [inline src/Uno.Foundation/UriExtensions.cs:60]: We should keep this internal for now.

File: src/Uno.UI/Controls/BindableImageView.Android.cs [hunk 3]
```diff
@@ -251,15 +252,23 @@ private void OnUriSourceChanged()
...
-				var filePath = UriSource.TrimStart("file://", StringComparison.OrdinalIgnoreCase);
+				string filePath;
+				if (newUri.IsAppData())
+				{
+					filePath = AppDataUriEvaluator.ToPath(newUri);
+				}
+				else
+				{
+					filePath = UriSource.TrimStart("file://", StringComparison.OrdinalIgnoreCase);
+				}
```
Comment (jeromelaban) [inline src/Uno.UI/Controls/BindableImageView.Android.cs:271]: ternary expression would be better.

**Comment by jeromelaban** [comment]
Looks like the build failed: https://dev.azure.com/uno-platform/Uno%20Platform/_build/results?buildId=13813&view=logs&j=8ba8221c-5c52-5996-1f75-e9a4f4aafd3b&t=1ad9a1d2-42b5-5738-5764-4f0f20081cba&l=5465

**Comment by jeromelaban** [comment]
Looks like it worked :) https://dev.azure.com/uno-platform/Uno%20Platform/_build/results?buildId=13846&view=ms.vss-test-web.build-test-results-tab&runId=1062988&resultId=100698&paneView=attachments

**Pull request merged**